src/loctool/PendoXliffFile.ts

```diff
@@ -0,0 +1,426 @@
+     * Given a target locale (as provided by loctool), return the locale that should be used
+     * in the output file.
+     */
+    private readonly getOuputLocale: (loctoolLocale: string) => string;
+
+    /**
+     * Create a deep copy of the supplied XLIFF object.
+     */
+    private static copyXliff(xliffDocument: Element): Element {
+        return xml2js(js2xml(xliffDocument, { compact: false }), { compact: false }) as Element;
+    }
+
+    /**
+     * Given a set of translations provided by loctool, for each locale
+     * write out a localized Pendo XLIFF file which is a copy of the source Pendo XLIFF but
+     * with applicable translated strings inserted.
+     */
+    localize(translations: TranslationSet, loctoolLocales: string[]): void {
+        if (!this.xliffDocument || !this.escapedUnits || !this.componentLists) {
+            throw new Error("Invalid operation: attempt to localize without extracting first.");
+        }
+
+        for (const loctoolLocale of loctoolLocales) {
+            const translationsForLocale = translations
+                .getAll()
+                .filter(
+                    (resource) => resource.getType() === "string" && resource.getTargetLocale() === loctoolLocale,
+                ) as ResourceString[];
+
+            // make a deep copy of the source xliff - don't mutate the file that's already loaded
+            const localizedXliffDocument = PendoXliffFile.copyXliff(this.xliffDocument);
+            const xliff = PendoXliffFile.getXliffRoot(localizedXliffDocument);
+
+            // apply the locale mapping so that the file has expected target locale set
+            const outputLocale = this.getOuputLocale(loctoolLocale);
+
+            // set target locale in all output files
+            // note: this is a mapped locale which can be different from the locale provided by loctool
+            const files = PendoXliffFile.findFileElements(xliff);
+            for (const file of files) {
+                file.attributes[XLIFF.ATTRIBUTE.TARGET_LOCALE] = outputLocale;
+            }
+
+            // mutate the translation units in the copy
+            const copyUnits = files.flatMap((file) => PendoXliffFile.findTransUnitElements(file));
+            for (const unitElement of copyUnits) {
+                try {
+                    // extract translation unit data for easier access
+                    const unitData = PendoXliffFile.extractUnitData(unitElement);
+
+                    // get translated string by matching TU ID
+                    const translation = translationsForLocale.find((resource) => resource.getKey() === unitData.key);
+                    if (!translation) {
+                        throw new Error("Missing translation for unit");
+                    }
+
+                    // get extracted component list based on TU key
+                    const referenceComponentList = this.componentLists[unitData.key];
+                    if (!referenceComponentList) {
```

Comment:
It would be empty. Your question made me realize there is no point in going through the whole backconversion logic if no components were extracted to begin with. I've updated the code in 57864af so that now `null` indicates no syntax was escaped, hence the backconversion can be sidestepped entirely.